### PR TITLE
Make restricted checkpoint loading thread-safe and 40% faster, simplify RLE prior

### PR DESCRIPTION
**Pull request by pderrenger**
## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

### 🌟 Summary
Restricted checkpoint loading now uses a locked, process-wide allow-list containing only checkpoint-referenced globals, while the RLE prior is computed directly as a standard normal log-probability.

### 📊 Key Changes
- Updated `_SafeLoad` to build and cache a registry of allowed globals, register checkpoint-specific entries with `add_safe_globals`, and protect registration with a process-wide lock.
- Kept restricted-load state thread-local and added a concurrent loading test covering 32 loads across eight threads.
- Restricted loading now depends on `get_unsafe_globals_in_checkpoint`; unsupported or older PyTorch versions continue to fall back to standard loading.
- Replaced `RealNVP.prior` and its `MultivariateNormal` construction with a closed-form two-dimensional standard-normal calculation, while retaining `loc` and `cov` buffers for checkpoint loading.
- Updated the package version from `8.4.125` to `8.4.126`.

### 🎯 Purpose & Impact
- Concurrent restricted loads no longer remove each other’s registered globals, and subsequent `weights_only=True` loads use the process-wide registrations.
- Restricted loading registers only globals referenced by each checkpoint, including supported torchvision transforms when present.
- RLE loss behavior is unchanged apart from the prior calculation path; non-finite error filtering remains in place.

### CODE REVIEW

**Comment by UltralyticsAssistant** [comment]
👋 Hello @pderrenger, thank you for submitting a `ultralytics/ultralytics` 🚀 PR! This automated message confirms your contribution was received, and an Ultralytics engineer will assist with the review. To ensure a seamless integration of your work, please review the following checklist:

- ✅ **Define a Purpose**: Clearly explain the purpose of your fix or feature in your PR description, and link to any [relevant issues](https://github.com/ultralytics/ultralytics/issues). Ensure your commit messages are clear, concise, and adhere to the project's conventions.
- ✅ **Synchronize with Source**: Confirm your PR is synchronized with the `ultralytics/ultralytics` `main` branch. If it's behind, update it by clicking the 'Update branch' button or by running `git pull` and `git merge main` locally.
- ✅ **Ensure CI Checks Pass**: Verify all Ultralytics [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI) checks are passing. If any checks fail, please address the issues.
- ✅ **Update Documentation**: Update the relevant [documentation](https://docs.ultralytics.com) for any new or modified features.
- ✅ **Add Tests**: If applicable, include or update tests to cover your changes, and confirm that all tests are passing.
- ✅ **Sign the CLA**: Please ensure you have signed our [Contributor License Agreement](https://docs.ultralytics.com/help/CLA) if this is your first Ultralytics PR by writing "I have read the CLA Document and I sign the CLA" in a new message.
- ✅ **Minimize Changes**: Limit your changes to the **minimum** necessary for your bug fix or feature addition. _"It is not daily increase but daily decrease, hack away the unessential. The closer to the source, the less wastage there is."_ — Bruce Lee

For more guidance, please refer to our [Contributing Guide](https://docs.ultralytics.com/help/contributing). Don't hesitate to leave a comment if you have any questions. Thank you for contributing to Ultralytics! 🚀

File: ultralytics/nn/tasks.py [hunk 1]
```diff
@@ -1612,27 +1612,20 @@ def restricted(cls):
...
-        if unsafe_globals is None or any(name.startswith("torchvision.transforms.") for name in unsafe_globals):
-            import torchvision.transforms.transforms as tvt
-            from torchvision.transforms.functional import InterpolationMode
-
-            allow = [*allow, tvt.Compose, tvt.Normalize, tvt.Resize, tvt.CenterCrop, tvt.ToTensor, InterpolationMode]
+            # Add the allow-list once per process: torch.serialization.safe_globals() removes its entries from a
+            # process-global set on exit, so with concurrent loads one thread's exit strips the allow-list out of
+            # another thread's in-flight unpickle.
+            allow = cls._build()
+            torch.serialization.add_safe_globals(allow)
```
Comment (UltralyticsAssistant) [inline ultralytics/nn/tasks.py]: 💡 **MEDIUM**: `add_safe_globals()` permanently mutates PyTorch's process-global weights-only allow-list rather than scoping it to this call.  After a documented per-call `torch_safe_load(..., safe_only=True)`, an unrelated `torch.load(..., weights_only=True)` in the same process will accept every entry returned by `_build()`, and calling `torch.serialization.clear_safe_globals()` later leaves `_globals` cached so subsequent restricted loads never re-register the list. Preserve per-call isolation with synchronized registration/reference counting, or otherwise detect and recover from external allow-list changes.

File: ultralytics/nn/tasks.py [hunk 2]
```diff
@@ -1682,6 +1675,8 @@ def _build(cls):
         import pkgutil
 
         import torch.nn.modules as torch_nn
+        import torchvision.transforms.transforms as tvt
```
Comment (UltralyticsAssistant) [inline ultralytics/nn/tasks.py]: 💡 **MEDIUM**: This makes every first restricted load import `torchvision`, including detection-only checkpoints that contain no torchvision transforms. If the installed torchvision package is unavailable or fails during import (for example, a broken/incompatible binary), `_build()` raises before `torch.load()` runs and `safe_only=True` loading fails even though the checkpoint does not need torchvision. Keep these transform globals lazy/conditional instead of importing them unconditionally while building the common allow-list.

**Review by UltralyticsAssistant — COMMENTED** [review]
## 🔍 PR Review

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

Reviewed all four changed files and the surrounding checkpoint-loading, model-construction, and RLE-loss paths. The RealNVP closed form and version bump are behaviorally sound, but the restricted-loader changes introduce process-global allow-list leakage/cache invalidation issues and an unconditional torchvision dependency on the safe-load path.

<details><summary>💬 Posted 2 inline comments</summary>

- 💡 **MEDIUM** `ultralytics/nn/tasks.py:1624` `add_safe_globals()` permanently mutates PyTorch's process-global weights-only allow-list rather than scoping it to this call.  After a documented per-call `torch_safe_load(..., safe_only=True)`, an unrelated `torch.load(..., weights_only=True)` in the same process will accept every entry returned by `_build()`, and calling `torch.serialization.clear_safe_globals()` later leaves `_globals` cached so subsequent restricted loads never re-register the list. Preserve per-call isolation with synchro…
- 💡 **MEDIUM** `ultralytics/nn/tasks.py:1678` This makes every first restricted load import `torchvision`, including detection-only checkpoints that contain no torchvision transforms. If the installed torchvision package is unavailable or fails during import (for example, a broken/incompatible binary), `_build()` raises before `torch.load()` runs and `safe_only=True` loading fails even though the checkpoint does not need torchvision. Keep these transform globals lazy/conditional instead of importing them unconditionally while building the…
</details>

File: ultralytics/nn/tasks.py (hunk 1 above)
Comment (pderrenger) [reply ultralytics/nn/tasks.py]: Addressed in 763dbf566 (and e04f9a340). Registration is deliberately process-permanent — any scoped removal is the cross-thread race this PR fixes — but it is no longer the whole `_build()` list: `loading()` registers only the globals `get_unsafe_globals_in_checkpoint` reports as currently missing for that checkpoint (YOLO26n registers 20, not 660), under a class lock. Because that scan runs on every load against torch's live set there is no cached sentinel any more: after `clear_safe_globals()` or a third-party `safe_globals()` scope exits, the next load re-reports and re-registers what it needs. The class docstring now states the process-wide scope explicitly.

File: ultralytics/nn/tasks.py (hunk 2 above)
Comment (pderrenger) [reply ultralytics/nn/tasks.py]: Addressed in 366fbe60a / e04f9a340: torchvision stays lazy exactly as #25883 made it. The transform entries are imported and added to the registry only when the checkpoint scan reports a `torchvision.transforms.*` global; detection-only checkpoints never import torchvision (verified: `'torchvision' in sys.modules` is False after a yolo26n restricted load, and `test_cli_imports_defer_torchvision` passes).

**Comment by UltralyticsAssistant** [comment]
🎉 [PR #25885](https://github.com/ultralytics/ultralytics/pull/25885) has been merged! Thank you @pderrenger, with contributions from @glenn-jocher, for strengthening restricted checkpoint loading and simplifying the RLE prior calculation.

> “Great things are done by a series of small things brought together.” — Vincent van Gogh

These changes improve concurrent loading safety with process-wide allow-list registration and locking, while preserving RLE behavior and updating the package to 8.4.126.
